src/main/java/com/android/volley/toolbox/ImageRequest.java

```diff
@@ -215,9 +219,21 @@ private static int getResizedDimension(int maxPrimary, int maxSecondary, int act
     }
 
     @Override
+    public void cancel() {
+        super.cancel();
+        synchronized (mLock) {
```

Comment:
That is possible, yes, but I don't see any problems it would cause, i.e. reasons this approach wouldn't meet the more tightly-defined guarantees I've noted in the Request#cancel Javadoc, or scenarios holding the same lock would fix.

Do you see something bad that can happen during this window?

Reply:
The contract is that the request is either live and the state is consistent, or the request is canceled and the state is cleared. You can't guarantee this using two disjoint locks.

Reply:
Whether the request is canceled or not is solely up to mCanceled. Other variables being mutated in cancel() have no bearing on the cancellation state. There is no guarantee that the listener is cleared at the same time as mCanceled, but I don't see why that matters.

Reply:
I'm just trying to make sense of the changes. I think it's confusing to have an identically-named lock `mLock` in the parent and in every subclass when they all essentially do the same thing. In ImageRequest, for example, if you're not going to share the lock, I think it might be more readable to remove the new lock and declare the listener volatile. That eliminates the red flags regarding deadlock as well as the casual reader's assumption that mLock is shared. However, my intuition is that the parent+subclass state should be managed by one lock.

Reply:
(I'm going to go ahead and merge this soon; if there are follow up concerns that require code changes, I can address them in a separate change).

The lock in each class is a private variable; they aren't considered part of the API of the class. Subclasses of Request don't have to worry about mLock at all. They only optionally have to worry about nulling their listener in cancel() if they care about preventing the leak, and doing so in a thread safe way to avoid an NPE in deliverResponse after a call to cancel(). They can use volatile if they like instead, or just not worry about nulling it at all and it'll work exactly the same as it does today.

Declaring the listener volatile doesn't change anything about thread-safety guarantees - it still would be modified outside of the lock held by Request - and I think it is debatable whether it is easier to read as volatile variables can be very tricky to reason about vs. a simple lock, and also not necessarily more performant especially if you have a bunch of them.

Let me try to lay out in more detail why it's fine for the locks to be separate. ExecutorDelivery checks if a request is canceled, and if not, it invokes deliverResponse/deliverError. This is inherently not thread-safe, because the request can be canceled after the check but before the invocation. So, there are two other things that can be done to make it thread-safe, which are now clearly documented in Request#cancel():

- Ensure cancel() is invoked on the same thread as ResponseDelivery, which is the main thread by default. In this case, the two will be serialized in some order, so explicit locking isn't needed.
- Have the custom listener null out the listener when overriding cancel() using its own thread-safe mechanism. In this case, it is guaranteed that once cancel() returns to the caller, the listeners will never be invoked, even if called from another thread. That is because both listeners are nulled out within some lock that is also grabbed when the listeners are read in deliverResponse/deliverError, although the lock is different in each case.

Now it's certainly possible that isCanceled() will return true to some other thread than the one  calling cancel(), even though the listener hasn't been nullified yet, if it is called before cancel() has returned. But this does not impact the guarantees above, and I don't think the benefits of offering such a guarantee are worth the API complexity for typical use cases of cancel.

Overall, the only purposes to nullifying the listener in subclasses are to:

- Mitigate a leak of the activity by dropping references to it when no longer needed. As long as the listener isn't invoked after being nullified, this can happen at any time after cancel().
- Try to offer a stronger guarantee about not invoking the listener after cancel() on a separate thread from the ResponseDelivery. Currently there is no guarantee at all. With this change, as long as you have ensured that cancel() has returned control to the caller, the listener won't be invoked.

The current approach should accomplish these goals, as best as I can tell.